### ISSUE TITLE
Better Methods for Fuzzy Join? 

### ISSUE THREAD

**Issue by EmmaLiTsai** [EmmaLiTsai/recoverKBTDR#16]
The centering functions (r_scripts/centering_functions.R) currently depend on the FuzzyJoin package to perform an incomplete merge (difference_left_join) with the trace and time dot data frames, such that the y-values of the time dot data frame can be used to center the scan. While this method works for my data, it is a bit slow and can crash if the record is 7-10 days long (data frame with ~500,000 rows). 

Is there a way to remove the dependency on this package and make this centering method more efficient? I was thinking this could possibly involve the caTools moving windows package, or hand-coding a "window" to center the records by.

**Comment by dschwilk** [comment]
Yes, I think this should be a straightforward assign trace points to dot according to window. If traces Xs are "x" and dot X values are "d", algorithm would be

For any dot i, assign trace points based on trace x values: dot i receives all trace points trace x > (d_(i-1) + d_i)/2 <= (d_i + d_(i+1)) / 2)

With special cases for dot 1 and final dot  -- use min(0, ) and max(n, )

I can give this a try soon once I clear some tasks on my todo list.

**Comment by dschwilk** [comment]
I'm working on this. Quick question. Why does the sample tracce file have four columns (X, Y, y_corr, y_corr_p2)? What are the last two?

**Comment by dschwilk** [comment]
OK, I have fast three line version. But it is hard to test against yours since yours did some things I don't understand. The original center trace deletes rows of data from the trace! WHY? You mentioned removing duplicates but shouldn't that happen on trace read-in or in the tidy function? I need to figure out how to test these.

**Comment by dschwilk** [comment]
Can you explain this? This looks like it should be a single mathematical expression. I don't understand

```
  fuzzy_merge_trace$y_val_corr <- dplyr::case_when(fuzzy_merge_trace$Y < 0 ~ (abs(fuzzy_merge_trace$Y) - dist_timedot), 
                                                   fuzzy_merge_trace$Y > 0 ~ ((-dist_timedot) - abs(fuzzy_merge_trace$Y)),
                                                   fuzzy_merge_trace$Y == 0 ~ (abs(fuzzy_merge_trace$Y) - dist_timedot))
```
There are no real cases! this is all the same, right? You are doing the exact same thing in each case, just obfuscating the code.  This boils down to
```
  fuzzy_merge_trace$y_val_corr <- -fuzzy_merge_trace$Y - dist_timedot)
```
Right?

**Comment by EmmaLiTsai** [comment]
Yes, your method makes much more sense! I had to remove duplicates in the fuzzy merge process because it added duplicated rows when a single point along the trace merged with 2+ time dots. Since removing duplicates only keeps the first duplicated row, this ended up creating a lag in the centering process that I wasn't super comfortable with and explains how the two centering methods aren't identical but visually similar. Your method is very elegant and I was able to debug it after resolving some of my file tidying issues-- I'll get to work on creating some extra tests to compare the two methods. 

Thank you so much for your help with this!
